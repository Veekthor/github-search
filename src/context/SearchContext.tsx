import React, { createContext, useState } from "react";
import { GitHubUser, ISearchedContext, ISearchProvider } from "../interfaces";

export const SearchedContext = createContext<ISearchedContext>({
  result: null,
  setResult: () => {},
  isLoading: false,
  setLoading: () => {},
});

//I got the idea for the Context Provider wrapper from: https://felixgerschau.com/react-typescript-context/
export const SearchedProvider: React.FC<ISearchProvider> = ({ children }) => {
  const [result, setResult] = useState<GitHubUser[] | null>(null);
  const [isLoading, setLoading] = useState<boolean>(false);

  return (
    <SearchedContext.Provider
      value={{
        setResult,
        result,
        isLoading,
        setLoading,
      }}
    >
      {children}
    </SearchedContext.Provider>
  );
};
