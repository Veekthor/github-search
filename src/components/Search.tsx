// TODO: RUN SEARCH IF URL IS LOADED WITH SEARCH TERM
import { useContext, useEffect, useState } from "react";
import { SearchedContext } from "../context/SearchContext";
import { IReturnedSearchCall } from "../interfaces";
import {
  SearchActions,
  SearchContainer,
  SearchInput,
} from "./styles/Search.styled";
import { useSearchParams } from "react-router-dom";

const Search: React.FC = () => {
  const [searchTerm, setTerm] = useState<string>("");
  const [searchType, setType] = useState<string>("user");
  const { setResult, isLoading, setLoading } = useContext(SearchedContext);
  const [params, setParams] = useSearchParams();

  const runSearch = async (
    term: string,
    type: string,
  ): Promise<IReturnedSearchCall> => {
    const url = `https://api.github.com/search/users?q=${term}+type:${type}`;
    const res = await fetch(url);
    const data = await res.json();
    return data;
  };

  const handleSearch = async (term: string, type: string) => {
    const response = await runSearch(term, type);
    setLoading(false);
    const fetchedData = response.items;
    setResult(fetchedData);
  };

  useEffect(() => {
    const term = params.get("term");
    const type = params.get("type");
    if (!term) return;
    setTerm(term);
    if (type) setType(type);
    setLoading(true);
    handleSearch(term, type || "user");
  }, []);

  const handleClick = async () => {
    setLoading(true);
    setParams({ term: searchTerm, type: searchType });
    handleSearch(searchTerm, searchType);
  };

  const handleChange = (e: React.ChangeEvent<HTMLInputElement>): void => {
    const { name, value } = e.target;
    if (name === "search") setTerm(value);
    if (name === "searchType") setType(value);
  };
  return (
    <SearchContainer>
      <SearchInput
        type="text"
        name="search"
        value={searchTerm}
        onChange={handleChange}
        data-testid="search-input"
      />
      <SearchActions>
        <div>
          <input
            type="radio"
            id="user"
            name="searchType"
            value="user"
            checked={searchType === "user"}
            onChange={handleChange}
          />
          <label htmlFor="user">User</label>
        </div>
        <div>
          <input
            type="radio"
            id="org"
            name="searchType"
            value="org"
            checked={searchType === "org"}
            onChange={handleChange}
          />
          <label htmlFor="org">Org</label>
        </div>
        <button
          onClick={handleClick}
          disabled={isLoading || !searchTerm.trim()}
        >
          Search
        </button>
      </SearchActions>
    </SearchContainer>
  );
};

export default Search;
