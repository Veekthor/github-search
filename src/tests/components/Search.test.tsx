import renderer from "react-test-renderer";
import Search from "../../components/Search";
import { Route, Routes, BrowserRouter as Router } from "react-router-dom";
import { act, render, screen, waitFor } from "@testing-library/react";
import { SearchedContext } from "../../context/SearchContext";
import userEvent from "@testing-library/user-event";
import fetchedUsersObj from "../fixtures/data";

const setLoading = jest.fn();
const setResult = jest.fn();

// Got idea for using context this way from https://polvara.me/posts/mocking-context-with-react-testing-library
const renderSearchWithContext = (isLoading: boolean) => {
  return render(
    <SearchedContext.Provider
      value={{
        result: [],
        isLoading,
        setLoading,
        setResult,
      }}
    >
      <Router>
        <Routes>
          <Route path="/" element={<Search />} />
        </Routes>
      </Router>
    </SearchedContext.Provider>
  );
};

describe("Search tests", () => {
  it("should match snapshot", () => {
    const component = renderer.create(
      <Router>
        <Routes>
          <Route path="/" element={<Search />} />
        </Routes>
      </Router>
    );
    const tree = component.toJSON();
    expect(tree).toMatchSnapshot();
  });

  it("should render all inputs in correct original state", () => {
    renderSearchWithContext(false);
    expect(screen.getByTestId("search-input")).not.toHaveValue();
    expect(screen.getByLabelText("User")).toBeChecked();
    expect(screen.getByLabelText("Org")).not.toBeChecked();
    expect(screen.getByText("Search")).toBeDisabled();
  });

  it("should disable only search button if in loading state", () => {
    renderSearchWithContext(true);
    const button = screen.getByText("Search");
    expect(button).toBeDisabled();
    expect(screen.getByTestId("search-input")).not.toBeDisabled();
    expect(screen.getByLabelText("User")).not.toBeDisabled();
    expect(screen.getByLabelText("Org")).not.toBeDisabled();
  });

  it("should correctly make API call and set states", async () => {
    fetchMock.resetMocks();
    const responseBody = fetchedUsersObj;
    fetchMock.mockResponseOnce(JSON.stringify(responseBody));
    act(() => {
      renderSearchWithContext(false);
    })
    expect(screen.getByText("Search")).toBeDisabled();
    const input = screen.getByTestId("search-input");
    act(() => {
      userEvent.type(input, "veek");
    })
    expect(screen.getByText("Search")).not.toBeDisabled();
    act(() => {
      userEvent.click(screen.getByText("Search"));
    })

    await waitFor(() => expect(setResult).toHaveBeenCalledTimes(1));
    expect(setLoading).toBeCalledTimes(2);
    expect(fetch).toBeCalledWith('https://api.github.com/search/users?q=veek+type:user')
    expect(setResult).toBeCalledWith(responseBody.items);
  })

  // TO DO: write test for api failures
});
