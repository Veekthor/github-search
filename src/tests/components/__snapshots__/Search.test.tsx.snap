// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Search tests should match snapshot 1`] = `
<section>
  <input
    aria-label="Search"
    className="sc-bgqQcB dLtZfJ"
    data-testid="search-input"
    name="search"
    onChange={[Function]}
    type="text"
    value=""
  />
  <div
    className="sc-gTRrQi bvvEia"
  >
    <div>
      <input
        checked={true}
        id="user"
        name="searchType"
        onChange={[Function]}
        type="radio"
        value="user"
      />
      <label
        htmlFor="user"
      >
        User
      </label>
    </div>
    <div>
      <input
        checked={false}
        id="org"
        name="searchType"
        onChange={[Function]}
        type="radio"
        value="org"
      />
      <label
        htmlFor="org"
      >
        Org
      </label>
    </div>
    <button
      disabled={true}
      onClick={[Function]}
    >
      Search
    </button>
  </div>
</section>
`;
